pkg/csi/service/vanilla/controller.go

```diff
@@ -1709,14 +1724,220 @@ func (c *controller) createFileVolume(ctx context.Context, req *csi.CreateVolume
+		if err != nil {
+			return nil, csifault.CSIInternalFault, logger.LogNewErrorCodef(log, codes.Internal,
+				"failed to get accessibility requirements by VC. Error: %+v", err)
+		}
+		log.Debugf("Topology accessibility requirements per VC are %+v", vcTopologySegmentsMap)
+	} else {
+		if topologyRequirement != nil {
+			// Get accessibility requirements.
+			for _, topology := range topologyRequirement.Preferred {
+				vcTopologySegmentsMap[c.managers.CnsConfig.Global.VCenterIP] = append(
+					vcTopologySegmentsMap[c.managers.CnsConfig.Global.VCenterIP],
+					topology.GetSegments())
+			}
+			log.Debugf("Topology accessibility requirements per VC are %+v", vcTopologySegmentsMap)
+		}
+	}
 	if !volTaskAlreadyRegistered {
 		var createVolumeSpec = common.CreateVolumeSpec{
 			CapacityMB: volSizeMB,
 			Name:       req.Name,
 			ScParams:   scParams,
 			VolumeType: common.FileVolumeType,
 		}
+		if topologyRequirement != nil {
+			var fsEnabledCandidateDatastores []*cnsvsphere.DatastoreInfo
+			var fsEnabledClusterToDsInfoMap map[string][]*cnsvsphere.DatastoreInfo
+			for vcHost := range vcTopologySegmentsMap {
+				// Get VC instance.
+				vcenter, err := common.GetVCenterFromVCHost(ctx, c.managers.VcenterManager, vcHost)
+				if err != nil {
+					return nil, csifault.CSIInternalFault, logger.LogNewErrorCodef(log, codes.Internal,
+						"failed to get vCenter instance for host %q. Error: %+v", vcHost, err)
+				}
+				// If Storage policy is given, check if it exists in the VC. If not found, continue to next VC.
+				var storagePolicyID string
+				if scParams.StoragePolicyName != "" {
+					storagePolicyID, err = vcenter.GetStoragePolicyIDByName(ctx, scParams.StoragePolicyName)
+					if err != nil {
+						// TODO: As govmomi doesn't support locale and all error messages are
+						// in English, we are temporarily resorting to a error message check.
+						// In future, we need to change govmomi to throw a NotFound error and catch that instead.
+						errMssgFromPBM := fmt.Sprintf("no pbm profile found with name: %q",
+							scParams.StoragePolicyName)
+						if err.Error() == errMssgFromPBM {
+							errMsg := fmt.Sprintf("Storage policy name %q not found in VC %q",
+								scParams.StoragePolicyName, vcenter.Config.Host)
+							log.Warn(errMsg)
+							continue
+						}
+						return nil, csifault.CSIInternalFault, logger.LogNewErrorCodef(log, codes.Internal,
+							"failed to get policy ID for storage policy name %q. Error: %+v",
+							scParams.StoragePolicyName, err)
+					}
+					log.Infof("Found ID %q for storage policy name %q in vCenter %q", storagePolicyID,
+						scParams.StoragePolicyName, vcenter.Config.Host)
+				}
+
+				var candidateDatastores []*cnsvsphere.DatastoreInfo
+				// Get accessible datastores in the matching topology requirement.
+				if commonco.ContainerOrchestratorUtility.IsFSSEnabled(ctx, common.TopologyPreferentialDatastores) {
```

Comment:
why are we checking preferential datastores FSS here?

Reply:
Yes, this can be removed. I think it got copied over from other existing logic

Reply:
Done